notebooks, scripts and models

NokeYuan opened the pull request:
inital commit for all codes

NokeYuan committed:
notebooks, scripts and models

inital commit for all codes

KevinHeUHN commented on env.yaml:
```diff
@@ -0,0 +1,553 @@
+  - zope.interface=5.4.0=py39h2bbff1b_0
+  - zstandard=0.19.0=py39h2bbff1b_0
+  - zstd=1.5.2=h19a0ad4_0
+  - pip:
```
Can you insert the contents of requirements.txt here under pip? And delete that file.

NokeYuan replied:
done

KevinHeUHN commented on scripts/ed_association.py:
```diff
@@ -0,0 +1,54 @@
+    test_df = pd.read_csv(test_df_path)
+    labels_list = [s for s in test_df.columns.tolist() if s.startswith("Label") and s.endswith("3pt_change")]
+
+    label_with_ed_per = process_results(results_df, test_df, labels_list, 'lgbm', 'Label_EDvisit')
```
add simple comments on what per and what all refers to

NokeYuan replied:
done

KevinHeUHN commented on scripts/train_pipeline.py:
```diff
@@ -0,0 +1,463 @@
+from src.utility import get_neg_to_pos_ratios
+
+
+def perform_bayesian_optimization(objective_func, model_name, model_type, pbounds, train_df, tune_df,test_df, label,full_colmuns_list, n_iter=3, init_points=5, weights=None, return_optimizer=False,verbose = False,allow_duplicate_points = True,ci =False):
```
When functions have a lot of arguments, make it more readable by splitting the arguments into separate lines
```
def perform_bayesian_optimization(
    objective_func, 
    model_name, 
    model_type,
    ...
):
```

NokeYuan replied:
done, all functions that have greater than 5 arguments will be splitted into separate lines

KevinHeUHN commented on scripts/train_pipeline.py:
```diff
@@ -0,0 +1,463 @@
+        nonlocal best_all_labels
+
+
+        objective_auroc, val_aucs ,auroc_ci, all_preds,all_labels,_,_ = objective_func(model_name, model_type, params, train_df, tune_df, test_df, label,full_colmuns_list, weights, verbose,ci)
```
arguments can also be split up into separate lines here too. In general, try not to go over 120 characters per line.

NokeYuan replied:
done

KevinHeUHN commented on scripts/train_pipeline.py:
```diff
@@ -0,0 +1,463 @@
+
+    if return_optimizer:
+        # If return_optimizer is True, return a tuple containing both the optimizer and best_val_aucs
+        return optimizer, best_val_aucs,best_auroc_ci,best_all_preds,best_all_labels
```
When returning multiple outputs, I think it would be better to just a return a dictionary 
```
return {
    'optimizer': optimizer,
    'best_val_aucs': best_val_aucs,
    ...
}
```

KevinHeUHN commented on scripts/train_pipeline.py:
```diff
@@ -0,0 +1,463 @@
+        # Otherwise, just return the best parameters as before
+        return optimizer.max['params']
+
+def train_model(model_name,model_type,params, train_df,tune_df,test_df,label,full_colmuns_list,weights=None,verbose = False,ci= False,testing = False):
```
same here

NokeYuan replied:
done

KevinHeUHN commented on scripts/train_pipeline.py:
```diff
@@ -0,0 +1,463 @@
+
+
+    if model_type == 'dl':
+        torch.cuda.empty_cache()
```
redundant code? torch.cuda.empty_cache is already called at the top

KevinHeUHN commented on scripts/train_pipeline.py:
```diff
@@ -0,0 +1,463 @@
+        if not isinstance(weights, list):
+            weights = [weights]
+
+        torch.cuda.empty_cache()
```
same here

KevinHeUHN commented on scripts/train_pipeline.py:
```diff
@@ -0,0 +1,463 @@
+            num_heads = {1: 8, 2: 16, 3: 32, 4: 64}
+            dim_feedforward = {1: 128, 2: 256, 3: 512}
+            hidden_size =  {1: 128, 2: 256, 3: 512}
+            model = Transformer(input_size, output_size,hidden_size =hidden_size[int(params['hidden_size'])],
```
nitpick but keep indentations consistent. Also try to clean up the spacing i.e. leave a space after a comma, don't have a space before =

KevinHeUHN commented on scripts/train_pipeline.py:
```diff
@@ -0,0 +1,463 @@
+        optimizer = optim.Adam(model.parameters(), lr=learning_rate, weight_decay=weight_decay)
+
+        if testing == True:
+            train_losses, train_aucs, train_aups, val_losses, val_aucs, val_aups, mean_val_losses, train_auc_cis, train_aup_cis, val_auc_cis, val_aup_cis,ir_models,models,all_preds,all_labels = train_and_evaluate_dl(
```
As mentioned before, return a dictionary of these output, and separate a long line of arguments into separate lines

KevinHeUHN commented on scripts/train_pipeline.py:
```diff
@@ -0,0 +1,463 @@
+            objective_auroc = val_aucs[0][best_epoch]  # find the epoch with the highest AUROC
+
+        else:
+            # Case when there are multiple outputs
```
No need to change anything now but why take the mean over all epochs? Instead of taking the highest auc across all epochs (for that particular target)?

The auc in the beginning would be bad since the model is just starting to learn.

NokeYuan replied:
`val_aucs` is a list of lists, initially designed for multi-task learning, where each sublist corresponds to the AUCs for a specific target. In train.py,` val_aucs` is defined as `val_aucs = [[] for _ in range(output_size)]`. If the output size is 1, this means` val_aucs` would have the structure` [[target1]]`, and `val_aucs[0] `would access the AUCs for this single target. Then, `best_epoch` selects the highest corresponding index of the list. 

In the multi-task scenarios, mean_aucs records the mean AUCs across 9 targets for each epoch. Then, using `np.argmax`, it finds the highest mean AUROC

KevinHeUHN replied:
Ohhh I see I see. Yea makes sense, thanks for that explanation!

KevinHeUHN commented on scripts/train_pipeline.py:
```diff
@@ -0,0 +1,463 @@
+            all_evaluations[(model, label)] = optimizer.res
+
+        # Multi-task (all labels) for DL models
+        for model in dl_models:
```
Is this supposed to be inside the loop? (Do this for each label?)

NokeYuan replied:
Should be outside of the loop since it does not require loop over each target.  Fixed.

KevinHeUHN commented on src/train.py:
```diff
@@ -0,0 +1,451 @@
+                        zip(train_aucs, train_aups, train_auc_cis, train_aup_cis, val_losses, val_aucs, val_aups,
+                            val_auc_cis, val_aup_cis)):
+                    print(
+                        f"Task {i + 1} {labels_list[i]}: Train AUC: {train_auc[-1]:.4f} (95% CI: {train_auc_ci[-1][0]:.4f}, {train_auc_ci[-1][1]:.4f}), Train AUP: {train_aup[-1]:.4f} (95% CI: {train_aup_ci[-1][0]:.4f}, {train_aup_ci[-1][1]:.4f}), Val Loss: {val_loss[-1]:.4f}, Val AUC: {val_auc[-1]:.4f} (95% CI: {val_auc_ci[-1][0]:.4f}, {val_auc_ci[-1][1]:.4f}), Val AUP: {val_aup[-1]:.4f} (95% CI: ...
```
Long print statements can also be split into separate lines

NokeYuan replied:
make sense, this is a print statament monitor the performance on the fly. Separate lines would casue taking up too much spaces on the screen, can comment out.

KevinHeUHN commented on src/train.py:
```diff
@@ -0,0 +1,451 @@
+        else:
+            ir_model_epoch = None
+
+        if my_copy == True:
```
What is my_copy for?

NokeYuan replied:
If copy is True, `copy.deepcopy(model)`  save the model(s) into list and return once finish training.

KevinHeUHN replied:
I see, so you have the option to store models for every epoch.

No need to change, but how I would do it is probably write checkpoint files for each epoch.

KevinHeUHN commented on src/utility.py:
Nice. I think this can be split into different files. One for plots, called viz.py. One for evaluation functions, called eval.py. And the rest we should put in util.py, so we don't have two different utility files.

KevinHeUHN commented:
Looks great! Only couple things to reorganize and nitpicks here and there.

KevinHeUHN reviewed (commented):
Looks great! Only couple things to reorganize and nitpicks here and there.

KevinHeUHN commented on src/train.py:
```diff
@@ -0,0 +1,451 @@
+    with torch.no_grad():
+        for i, batch in enumerate(dataloader):
+            inputs, labels = batch
+            inputs, labels = (inputs.to(device, dtype=torch.float), labels.to(device, dtype=torch.float))
```
Also, try to minimize repeated code. validate_dl has the same block of code as here. You can place that block of code into a predict function.

KevinHeUHN reviewed (commented):
Looks great! Only couple things to reorganize and nitpicks here and there.

NokeYuan committed:
updated code

merge code into util.py

NokeYuan committed:
update analysis

NokeYuan committed:
update analysis

NokeYuan committed:
remove utility.py

merge it into util.py

NokeYuan committed:
create eval and viz

NokeYuan committed:
update comment

NokeYuan committed:
Merge branch 'main' into noke_dev

KevinJCHe committed:
Update util.py

KevinJCHe commented on src/util.py:
```diff
@@ -4,8 +4,21 @@
+# Standard library imports
+import os
+from functools import reduce
+os.environ['OUTDATED_IGNORE'] = '1'
```
Just curious, what is this for?

NokeYuan replied:
it will ignore the outdated package reminder. E.g., xx latest version is 2.0.0 but your current is 1.5.3

KevinJCHe replied:
Ahh cool thanks!

NokeYuan replied:
os.environ['OUTDATED_IGNORE'] = '1' would ignore message indicate which packages is outdated. E.g., xxx latest version is 2.0.0, your current version is 1.5.2

KevinJCHe commented on src/util.py:
```diff
@@ -87,4 +100,237 @@ def get_excluded_numbers(df, mask: pd.Series, context: str = '.') -> None:
+    largest_interval = max(interval_list, key=lambda x: x[1]-x[0])
+    return largest_interval
+
+def partial_predict(X,label,model_name,train_results,df,device):
```
Leave a quick docstring here explaining this is used for computing SHAP values for models taking sequential input (e.g. recurrent neural network)

NokeYuan replied:
done

KevinJCHe reviewed (changes requested):
Yea nice job on creating the eval.py and viz.py!

Nothing too much to add from last time. Would you able to fix up the previous comments I left? 

Then we can merge this branch

NokeYuan committed:
Adjustment

1. Add Rscripts for boostraping and GLMM model
2. Revise styling to seperate arguments as lines based on comments.
3. Add doc string for function partial_predict

NokeYuan committed:
Merge branch 'noke_dev' of github.com:ml4oncology/PredUCE into noke_dev

# Conflicts:
#	requirements.txt
